File: scala/sieve/src/main/scala/Sieve.scala
object Sieve {
  def sieve(stream: Stream[Int]): Stream[Int] =
    stream.head #:: sieve(stream.tail.filter(_ % stream.head != 0))

  def primesUpTo(limit: Int) = sieve(Stream.from(2)).takeWhile(_ <= limit)
}

File: haskell/sieve/src/Sieve.hs
module Sieve (primesUpTo) where

primesUpTo :: Integral a => a -> [a]
primesUpTo limit = sieve [2..limit]
  where
    sieve [] = []
    sieve (x:xs) = x : sieve (filter ((/=0) . (`mod` x)) xs)